Feature node

jeremylfchen opened the pull request:
Hi guys, heres my first PR lol

Main summary of the PR:
- implement knex seeding files
- implement information generation in `data-gen-functions.js`
- implement csv loading script in `data-gen/index.js`
- test file for information generation in `data-gen.test`

No need to pay attention to the following files:
- anything with `mongo` in its file path (commented out the old seed code to speed up testing
- client.test
- server.test

jeremylfchen committed:
house keeping

created database folder for all database code

jeremylfchen committed:
start data-gen script

jeremylfchen committed:
completed data gen script
export data to csv

jeremylfchen committed:
fixed item id to be for 200 - 10M+199

jeremylfchen committed:
knex migration file wip

jeremylfchen committed:
generate item id and make csv

jeremylfchen committed:
refactor seed script

include ids for stores and items tables

jeremylfchen committed:
refactor data generating scripts
- information generating functions now exist in data-gen-functions
- actual csv filling opertion done in index to prevent accidental re-seeding of csv files during testing

complete testing for csv populating and data generation

Nick-Zabalza commented:
@jeremylfchen 

Hi Jeremy,

For the most part, looks good. I definitely appreciate that the test file doesn't require 10,000,000 records to be generated for it to run properly.

There did seem to be some issues though.

1. Looking at the data generation script, the itemMaker function looks like it will be able to produce items with the correct item ID starting at 100, since you add startIdx to whatever is passed into it. But I did not see the same being done in storeMaker and availabilityMaker. Is this intended? The way I see it, both the startIdx and having to minus 1 from id + startIdx (in itemMaker) shouldn't be necessary. Instead, I feel it would be more efficient for dataWriter to produce the correct itemId, starting at 100 and then just pass that to the three Maker functions. What do you think?

2. Due to number 1, after I ran the tests, I checked the CSV files  and itemMaker does correctly produce files starting at itemId 100 while availablityMaker looks to be starting itemId at 1. This doesn't seem to be intended. Further, how were your tests able to mostly pass with this issue intact?

3. Yes, I said tests "mostly pass". I got this error for your seed tests:
 ● dataWriter should:  › for items  › write 20 entries to a csv file

    expect(received).toEqual(expected) // deep equality

    Expected: "100"
    Received: "101"

      69 |           .pipe(csvParser())
      70 |           .on('data', (row) => {
    > 71 |             expect(row.item_id).toEqual('100');
         |                                 ^
      72 |           })
      73 |           .on('end', () => {
      74 |             readStream.destroy();

      at CsvParser.<anonymous> (test/data-gen.test.js:71:33)
      at CsvParser.writeRow (node_modules/csv-parser/index.js:191:10)
      at CsvParser.parseLine (node_modules/csv-parser/index.js:166:23)
      at CsvParser._transform (node_modules/csv-parser/index.js:251:16)


4. Lastly, your tests didn't run at first since there was no test-csv folder in your test folder. My understanding of GIT is that: even if you have a folder physically in your directory, and even if that folder is not being git.ignored, git still won't commit it if there is no actual comittable file inside it. So I would suggest either adding a dummy file to test-csv so that the folder exists for anyone using your repo or design your test script to check for that folder and, if it doesn't exist, create it.

You're almost there for this PR merge.

Nick

jeremylfchen committed:
updated variable names to reduce confusion

availability only stores the indexs of the item and stores, not the actual ID
- the reason for this is because item_ids start at 100 but SQL primary key indexs start at 1
- this might not make sense for store ids since it is an arbitrary number only used in my service, and I could technically omit this field and just use the primary key index as the store index. However I will keep it this way as a way to future proof my stores tables in the chance that store ids might one day follow a different numbering system (i.e. start at 100)

refactor data writer tests to output 2 csv files each
-  avoid overwriting data from first test

jeremylfchen commented:
Hi @Nick-Zabalza ,

> Looking at the data generation script, the itemMaker function looks like it will be able to produce items with the correct  item ID starting at 100, since you add startIdx to whatever is passed into it. But I did not see the same being done in storeMaker and availabilityMaker. Is this intended? 
Correct, sorry for the confusion. The availability table actually only stores the primary key index of the store and the item. And since the indexes start at 1 in SQL, that is why i start at that instead of 100.

As for the store table, store id naming is pretty arbitrary and I technically could just only have the primary key idx as the store id but I am keeping it the way it is as a way to "future proof" that table incase one day the store ids need to follow a different numbering system that doesnt start at 1. That way I can utilize table indexing for better look ups(?, this is based on my research, will need to test it out)


> The way I see it, both the startIdx and having to minus 1 from id + startIdx (in itemMaker) shouldn't be necessary. Instead, I feel it would be more efficient for dataWriter to produce the correct itemId, starting at 100 and then just pass that to the three Maker functions. What do you think?
For store id that could work, but since store id is pretty arbitrary Ill keep it the way it is.


As for the error in my tests, I didnt run into that problem. I refactored them to produce 2 test csvs per test. Looks like it failed the `for items: write an single entry to a csv file`. 
If you have time, could you run the tests again and share a screenshot of the contents of `itemTest1.csv` please?

Nick-Zabalza commented:
@jeremylfchen 
Now all the tests are passing. As for the IDs, i do not believe i am talking about the store Ids but the item Ids. It is the item Ids that are only be correctly assigned by one maker function. As you can see in the screen shot below, the item-idd starts at 1 for the availaility table

![image](https://user-images.githubusercontent.com/59750779/91796928-343a8300-ebd6-11ea-958d-49708e730117.png)

Nick
